### PR TITLE
Credentials domain pagination

### PR DESCRIPTION
**Pull request by johanbrandhorst**
Adds the logic for pagination in the credential domain. The handler logic will be reviewed in #3902. See that PR for all tests passing. This will have both build errors and failed tests.

### CODE REVIEW

File: internal/credential/service_list_credentials.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/hashicorp/boundary/internal/pagination"
+)
+
+// CredentialService defines the interface expected
+// to get the total number of credentials and deleted ids.
+type CredentialService interface {
+	EstimatedCredentialCount(context.Context) (int, error)
+	ListDeletedCredentialIds(context.Context, time.Time) ([]string, time.Time, error)
+	ListCredentials(context.Context, string, ...Option) ([]Static, error)
+}
```
Comment (johanbrandhorst) [inline internal/credential/service_list_credentials.go:20]: The credential service is implemented by any credential repos, in this case the only one is the static credential repo.

File: internal/credential/service_list_libraries.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+// LibraryService defines the interface expected
+// to list, estimate the count of and list deleted items of
+// credential libraries.
+type LibraryService interface {
+	EstimatedCount(context.Context) (int, error)
+	ListDeletedIds(context.Context, time.Time) ([]string, time.Time, error)
+	List(context.Context, string, ...Option) ([]Library, error)
+}
```
Comment (johanbrandhorst) [inline internal/credential/service_list_libraries.go:21]: The library service is implemented by any credential library repos, in this case just the vault repo.

File: internal/credential/service_list_stores.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/hashicorp/boundary/internal/util"
+)
+
+// SubtypeStoreService defines the interface expected
+// to gather information about credential stores.
+type SubtypeStoreService interface {
+	EstimatedStoreCount(context.Context) (int, error)
+	ListDeletedStoreIds(context.Context, time.Time, ...Option) ([]string, error)
+	ListCredentialStores(context.Context, []string, ...Option) ([]Store, error)
+}
```
Comment (johanbrandhorst) [inline internal/credential/service_list_stores.go:23]: The SubtypeStoreService is implemented by both the static and vault credential repos.

File: internal/credential/service_list_stores.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+// StoreService coordinates calls across different subtype services
+// to gather information about all credential stores.
+type StoreService struct {
+	services []SubtypeStoreService
+	writer   db.Writer
+}
+
+// NewStoreService returns a new credential store service.
+func NewStoreService(ctx context.Context, writer db.Writer, vaultService SubtypeStoreService, staticService SubtypeStoreService) (*StoreService, error) {
```
Comment (johanbrandhorst) [inline internal/credential/service_list_stores.go:33]: In theory we could take a slice of services, but I think it's more error prone than having two explicit parameters.

File: internal/credential/static/options.go [hunk 5]
```diff
@@ -64,3 +69,11 @@ func WithPrivateKeyPassphrase(with []byte) Option {
 		o.withPrivateKeyPassphrase = with
 	}
 }
+
+// WithStartPageAfterItem is used to paginate over the results.
+// The next page will start after the provided item.
+func WithStartPageAfterItem(item pagination.Item) Option {
+	return func(o *options) {
+		o.withStartPageAfterItem = item
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/credential/static/options.go:77]: This applies to this option in all pkgs.  I'm not going to keep repeating the
same question.

File: internal/credential/vault/service_library.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/util"
+)
+
+// LibraryService coordinates calls to gather information about all credential libraries.
+type LibraryService struct {
```
Comment (jimlambrt) [inline internal/credential/vault/service_library.go:15]: LibraryListingService?

File: internal/credential/vault/service_library.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	repo   *Repository
```
Comment (jimlambrt) [inline internal/credential/vault/service_library.go:16]: do we want to define an interface which constrains the dependency the receiver
funcs required?

File: internal/credential/vault/service_list.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+// List lists both generic and SSH credential libraries.
+// Supports the following options:
+//   - credential.WithLimit
+//   - credential.WithStartPageAfterItem
+func (s *LibraryService) List(ctx context.Context, credentialStoreId string, opts ...credential.Option) ([]credential.Library, error) {
+	const op = "vault.(*LibraryService).List"
+	opt, err := credential.GetOpts(opts...)
```
Comment (jimlambrt) [inline internal/credential/vault/service_list.go:22]: I'd prefer to see the parameters validated vs relying on other funcs to validate
them. (this preference applies to all the receiver funcs in this file I believe)

File: internal/credential/service_list_credentials.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/pagination"
+)
+
+// CredentialService defines the interface expected
+// to get the total number of credentials and deleted ids.
+type CredentialService interface {
```
Comment (jimlambrt) [inline internal/credential/service_list_credentials.go:16]: CredentialListingService?

File: internal/credential/service_list_libraries.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/pagination"
+)
+
+// LibraryService defines the interface expected
+// to list, estimate the count of and list deleted items of
+// credential libraries.
+type LibraryService interface {
```
Comment (jimlambrt) [inline internal/credential/service_list_libraries.go:17]: LibraryListingService?

File: internal/credential/service_list_stores.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/pagination"
+	"github.com/hashicorp/boundary/internal/util"
+)
+
+// SubtypeStoreService defines the interface expected
+// to gather information about credential stores.
+type SubtypeStoreService interface {
```
Comment (jimlambrt) [inline internal/credential/service_list_stores.go:19]: SubtypeStoreListingService?

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Just a few minor questions.  Looks good.  I'll be interested to see the outcome of today's meeting about the multiple table pagination queries.

File: internal/credential/vault/service_library.go (hunk 7 above)
Comment (johanbrandhorst) [reply internal/credential/vault/service_library.go:16]: I don't know that it's that useful when the expected type is in the same package. Better to test with a real repository than using a fake/mock, right?

File: internal/credential/vault/service_list.go (hunk 8 above)
Comment (johanbrandhorst) [reply internal/credential/vault/service_list.go:22]: Good shout, I'll add validation.

File: internal/credential/static/options.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/credential/static/options.go:77]: I'd be happy to add such a function, but not all options allow for validation logic when the option is used (see this one for example), how would you prefer validation to be done in those cases? This code is following the pattern in this file, which may or may not be ideal.

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/credential/static/options.go:77]: I added this function and called it in the places where we can reasonably perform validation of the item (the credential level options).

**Pull request closed without merging**